lib/DBus/Generation.hs

```diff
@@ -0,0 +1,539 @@
+        buildOutputNames = mapM (newNameDef . I.methodArgName) outputArgs
+        takeBusArg = isNothing busNameM
+        takeObjectPathArg = isNothing objectPathM
+        functionNameFirst:functionNameRest = coerce methodNameMN
+        functionName = (Char.toLower functionNameFirst):functionNameRest
+        functionN = mkName $ (Char.toLower functionNameFirst):functionNameRest
+        methodCallDefN = mkName $ functionName ++ "MethodCall"
+        defObjectPath = fromMaybe (fromString "/") objectPathM
+    clientN <- newName "client"
+    busN <- newName "busName"
+    objectPathN <- newName "objectPath"
+    methodCallN <- newName "methodCall"
+    callResultN <- newName "callResult"
+    replySuccessN <- newName "replySuccess"
+    methodArgNames <- buildArgNames
+    fromVariantOutputNames <- buildOutputNames
+    finalOutputNames <- buildOutputNames
+    let variantListExp = map makeToVariantApp methodArgNames
+        mapOrHead' = mapOrHead outputLength
+        fromVariantExp = mapOrHead' makeFromVariantApp fromVariantOutputNames TupE
+        finalResultTuple = mapOrHead' VarE finalOutputNames TupE
+        maybeExtractionPattern = mapOrHead' makeJustPattern finalOutputNames TupP
+        getMethodCallDefDec = [d|
+               $( varP methodCallDefN ) =
+                 M.MethodCall { M.methodCallPath = defObjectPath
+                              , M.methodCallInterface = Just methodInterface
+                              , M.methodCallMember = methodNameMN
+                              , M.methodCallDestination = busNameM
+                              , M.methodCallSender = Nothing
+                              , M.methodCallReplyExpected = True
+                              , M.methodCallAutoStart = True
+                              , M.methodCallBody = []
+                              }
+                 |]
+        setMethodCallParamsE = getSetMethodCallParams methodCallDefN
+                               (maybeName busN takeBusArg)
+                               (maybeName objectPathN takeObjectPathArg)
+                               (return $ ListE variantListExp)
+        handleReplySuccess =
+          if outputLength == 0
+          then
+            [| Right () |]
+          else
+            [|
+               case (M.methodReturnBody $( varE replySuccessN )) of
+                     $( return $ ListP $ map VarP fromVariantOutputNames ) ->
+                       case $( return $ fromVariantExp ) of
+                         $( return maybeExtractionPattern ) -> Right $( return finalResultTuple )
+                         _ -> Left $ clientArgumentUnpackingError $
+                              M.methodReturnBody $( varE replySuccessN )
+                     _ -> Left $ clientArgumentUnpackingError $
+                          M.methodReturnBody $( varE replySuccessN )
+             |]
+        getFunctionBody = [|
+             do
+               let $( varP methodCallN ) = $( setMethodCallParamsE )
```

Comment:
There are multiple warnings like this:
    This pattern-binding binds no variables:
      $(varP methodCallN) = $setMethodCallParamsE

I'm not familiar with template haskell, is this compiler misunderstanding something or a genuine issue?

Reply:
Yes, this is a known issue. See this thread https://mail.haskell.org/pipermail/haskell-cafe/2014-December/117437.html